tests/test_util/display.rs

```diff
@@ -0,0 +1,166 @@
+use chalk_integration::{program::Program, query::LoweringDatabase, tls};
+use chalk_solve::display::{self, WriterState};
+use regex::Regex;
+use std::{fmt::Debug, sync::Arc};
+
+macro_rules! reparse_test {
+    (program $program:tt) => {
+        crate::test_util::display::reparse_test(crate::test_util::strip_leading_trailing_braces(
+            stringify!($program),
+        ))
+    };
+    (program $program:tt produces $diff:tt) => {
+        crate::test_util::display::reparse_into_different_test(
+            crate::test_util::strip_leading_trailing_braces(stringify!($program)),
+            crate::test_util::strip_leading_trailing_braces(stringify!($diff)),
+        )
+    };
+    (program $program:tt formatting matches $res:literal) => {
+        crate::test_util::display::test_formatting(
+            crate::test_util::strip_leading_trailing_braces(stringify!($program)),
+            $res,
+        )
+    };
+}
+
+pub fn write_program(program: &Program) -> String {
```

Comment:
This looks pretty similar to the `write_program` function in `chalk-solve/display`?

Reply:
It is similar - good catch, we missed that one. One operates on the interface provided by `Program`, the other `RustIrLoggingDb`. I think we can write a function that transforms the data in a `Program` into a list of `RecordedItemIds` similar to what we have in the logging db, and then provide that to `chalk-solve::display::write_program`, eliminating the duplication of the writing logic. I'll add this to the pull request task list.